[hw/ip] OTP ROM patch generation tool

sameo opened the pull request:
This PR is a first attempt at building an OTP ROM patch generation tool, based on the proposed ROM patch specification changes tracked at #21070 

TODO before this PR can be merged:

* [x] Add tests to validate the generated patches
* [x] Define multiple output (OTP HJSON partition, binary blob)
* [x] Add ROM patch signature support

engdoreis commented on third_party/rust/Cargo.toml:
```diff
@@ -55,8 +55,8 @@ rsa = "0.8"
 serde_derive = "1.0"
-
 serde_bytes = "0.11"
+serde-hjson = "0.9"
```
@pamaury have more context on this, but I think that we had issues trying to parse some hjson files in the repo with the rust implementation. So unless we really need the hjson features, here I would suggest using json.

jwnrt replied:
@engdoreis is right, but if you _do_ need to use hjson then we found the [deser-hjson](https://crates.io/crates/deser-hjson) crate to have fewer bugs (but not zero) than serde-hjson.

pamaury replied:
Unless we want to serialize hjson, I would indeed suggest that we stick `deser-hjson` which we already use and which is working well (we successfully submitted upstream bug fixes).

sameo replied:
This is fixed with the updated PR.

engdoreis commented on sw/host/opentitantool/src/command/rom_patch.rs:
```diff
@@ -0,0 +1,84 @@
+        let mut buffer = vec![0_u8; self.otp_size];
+        _ = patch_partition.read(&mut buffer)?;
+
+        println!("uint8_t fake_patch[] =");
```
We have other similar tooling in the repo to generate code from a template + hjson file using the python library [mako templates](https://www.makotemplates.org/).
Is there a reason for the tool to be inside opentitantool or can this also be a python script using a template file?
In case it needs to be inside opentitantool, can we use a template crate?

sameo replied:
We are now generating an HJSON OTP partition image, to be used as an OTP image overlay.

engdoreis reviewed (commented):
Thanks @sameo for the PR, I please check the questions left below.

andreaskurth commented:
Thanks for this work and PR, @sameo! Let's get more reviews on this, so we can eventually be confident with merging it :-)

I'd like to review the first part you mention, "A ROM patch specification change", but couldn't find it in the files changed in this PR. Could you please point me to it?

jwnrt commented on sw/host/opentitanlib/src/rom_patch/mod.rs:
```diff
@@ -0,0 +1,7 @@
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// TODO(lowRISC/opentitan#15443): Fix this lint.
+#[allow(clippy::module_inception)]
```
I think you can remove these lines - the parent module is named `rom_patch` and the child is just `patch`, so there shouldn't be a warning.

Also, nit: do you need this `patch` submodule? Are you planning to add more modules to `rom_patch` in the future?

sameo replied:
I removed that comment. And I also removed `patch.rs` to keep all the code under `mod.rs`.

jwnrt commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,738 @@
+use std::fs;
+use std::io::{self, Read};
+use std::ops::Range;
+use std::path::Path;
+
+use std::{mem, ptr};
```
Nit:

```suggestion
use std::path::Path;
use std::mem;
use std::ptr;
```

sameo replied:
Fixed as well, thanks.

jwnrt commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,738 @@
+// The code section offset from the patch start
+const PATCH_CODE_SECTION_OFFSET: usize = mem::size_of::<RomPatchHeader>() + PATCH_MATCH_TABLE_SIZE;
+
+#[allow(dead_code)]
+enum MultiBitBool {
+    Bool4True = 0x6,
+    Bool4False = 0x9,
+}
```
Nit:

These constants are already available from [`crate::chip::boolean`](https://opentitan.org/gen/rustdoc/opentitanlib/chip/boolean/struct.MultiBitBool4.html) as `MultiBitBool4::True as u8` and `MultiBitBool4::False as u8`.

sameo replied:
Yep. I switched to the internal `MultiBitBool4` crate with the new PR.

jwnrt commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,738 @@
+    revision: u8,
+}
+
+impl Read for RomPatchHeader {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let raw_header = self as *const _ as *const u8;
+        let header_size = mem::size_of::<RomPatchHeader>();
+
+        if buf.len() < header_size {
+            return Err(std::io::Error::from(std::io::ErrorKind::InvalidInput));
+        }
+
+        unsafe {
+            ptr::copy_nonoverlapping(raw_header, buf.as_mut_ptr(), header_size);
+        }
+
+        Ok(header_size)
+    }
+}
```
Per @nbdd0121 - this probably shouldn't be an unsafe `Read` implementation, you can use [`zerocopy::AsBytes`](https://docs.rs/zerocopy/0.7.32/zerocopy/trait.AsBytes.html) instead. This is already in our dependency tree.

sameo replied:
All the code has been moved to use the `zerocopy` crate.

jwnrt commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,738 @@
+
+#[derive(Default, Clone, Copy, Debug)]
+#[allow(dead_code)]
+#[repr(packed)]
```
This needs to be `#[repr(packed, C)]` otherwise it will be `#[repr(packed, Rust)]` and rust can re-order the fields.

sameo replied:
Fixed as well, thanks.

jwnrt commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,738 @@
+
+impl fmt::Display for RomPatchHeader {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let unaligned = std::ptr::addr_of!(self.size);
+        let size = unsafe { std::ptr::read_unaligned(unaligned) };
+        write!(
+            f,
+            "\tMeta {:#x} Revision {:?} Size {:?} DWORDs",
+            self.meta, self.revision, size
+        )
```
```suggestion
        write!(
            f,
            "\tMeta {:#x} Revision {:?} Size {:?} DWORDs",
            self.meta, self.revision, {self.size}
        )
```

sameo replied:
This part of the code has been reworked. We should not do unaligned accesses anymore, except in the unit tests.

jwnrt commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,738 @@
+    }
+}
+
+impl Read for RomPatchSubRoutine {
```
This can also be a `zerocopy::AsBytes` call I think

sameo replied:
Indeed. We no longer use the `Read` trait.

jwnrt commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,738 @@
+            .to_le_bytes()[..];
+
+        /* Copy the signature */
+        unsafe {
+            ptr::copy_nonoverlapping(
+                signature.as_ptr(),
+                buf[self.signature_offset()..].as_mut_ptr(),
+                RSA_SIGNATURE_LEN_BYTES,
+            );
+        }
```
`signature` is already a `&[u8]` so I don't think you need to copy the memory like this.

sameo replied:
I switched to ECDSA signatures, and that has been fixed now.

jwnrt reviewed (changes requested):
Thanks @sameo - some comments on the unsafe code

alees24 reviewed (commented):
Thanks for this. The documentation/specification changes look good, modulo a few corrections and clarifications that I've suggested.
May I suggest, given the time frames involved, that it would be sensible to separate these from the patch generation tool so that the documentation and specification changes may be integrated sooner?

The lowRISC software team will want to review the tool and Rust code thoroughly and I think our priority should be to get the ROMs and e2e testing completed first.

pamaury commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,745 @@
+    Bool4False = 0x9,
+}
+
+#[derive(Deserialize, Debug)]
```
Unless I have missed it, the hjson format is not really documented. There is a (quite clear) example but I think it would be worth either having a proper document explaining the format, or having documentation comments in the rust code.

pamaury commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,745 @@
+
+#[derive(Copy, Clone, Debug, Default)]
+#[allow(dead_code)]
+#[repr(packed)]
```
Should this be `repr(packed, C)`?

pamaury commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,745 @@
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+#[allow(dead_code)]
```
Why is this necessary?

sameo replied:
Not relevant anymore with the update PR.

pamaury commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,745 @@
+    fn from_hjson(remap_base: u32, region: &RomPatchRegionHjson) -> Result<Self> {
+        let match_addr = u32::from_str(&region.match_addr)?;
+
+        let size_dwords = (region.patch_code.len() as u32) & 0xf;
```
should the code make sure that `patch_code` is not too long?

sameo replied:
That part has been removed from the PR.

pamaury commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,745 @@
+        let size_dwords = (region.patch_code.len() as u32) & 0xf;
+        let locked = if region.locked { 1 << 31 } else { 0 };
+
+        let match_address = match_addr & 0x7ffffff;
```
what happens if match_addr has its high bit set? Is that an error?

sameo replied:
I added a check for that. It should be an error indeed.

pamaury commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,745 @@
+        (self.match_base >> 31) != 0
+    }
+
+    fn size(&self) -> u8 {
```
I suggest renaming `size` to something more precise, like `size_dwords` (which I think is what it is here?)

sameo replied:
Not relevant anymore with the updated PR.

pamaury commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,745 @@
+        })
+    }
+
+    fn len(&self) -> usize {
```
I would suggest to rename `len` to something more precise, like `size_bytes`.

sameo replied:
ditto.

pamaury commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,745 @@
+
+struct RomPatch<'a> {
+    header: RomPatchHeader,
+    size: usize,
```
maybe document what this is the size of? And size in what unit (bytes?)

sameo replied:
Removed from the updated PR.

pamaury commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,745 @@
+        for (index, region) in patch.regions.iter().enumerate() {
+            let match_pair = RomPatchMatchDescriptor::from_hjson(remap_addr, region)?;
+            match_table[index] = match_pair;
+            remap_addr += (region.patch_code.len() as u32) * 4;
```
replace 4 by `std::size_of<u32>()`? Apply throughout.

sameo replied:
Yes. I have done that throughout all the code.

pamaury commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,745 @@
+        self.signature_offset
+    }
+
+    fn code_len(&self) -> usize {
```
rename to specify the unit

sameo replied:
Not relevant anymore.

pamaury commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,745 @@
+        self.signature_offset() - self.code_offset()
+    }
+
+    fn len(&self) -> usize {
```
what is this the length of?

sameo replied:
Ditto.

pamaury commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,745 @@
+        self.size
+    }
+
+    fn validate(&self) -> Result<()> {
```
what is the purpose of this method?

sameo replied:
I removed it.

pamaury commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,745 @@
+}
+
+impl<'a> Read for RomPatch<'a> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
```
As @jwnrt said above, this could use `zerobytes::FromBytes` to make it safer and clearer. See an example here:
https://github.com/lowRISC/opentitan/pull/20672/files#diff-4455f74af62f7ea7ec84c7fd2e934a0b14e940e9d917ae899d276fc5970968afR315

sameo replied:
Yes. The updated PR is now based on the `zerocopy` crate.

pamaury commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,745 @@
+}
+
+pub struct RomPatchPartition<'a> {
+    size: usize,
```
size of what?

sameo replied:
Not relevant anymore with the updated PR.

pamaury commented on sw/host/opentitantool/src/command/rom_patch.rs:
```diff
@@ -0,0 +1,84 @@
+/// Generate an OTP ROM patch partition
+#[derive(Debug, StructOpt)]
+pub struct Generate {
+    #[structopt(
```
Please use `clap`, we migrated away from `structopt` (which is the predecessor of structopt)

sameo replied:
I had to pull the `clap` crate update and the migration code, in order to proceed with that conversion.

sameo replied:
Done.

pamaury commented on sw/host/opentitantool/src/command/rom_patch.rs:
```diff
@@ -0,0 +1,84 @@
+}
+
+impl CommandDispatch for Generate {
+    fn run(
```
Can you explain/document what this command is supposed to generate? I was expecting it to generate a binary file (i.e. one that corresponds to the format that you implemented `Read` for). But this seems to generate a C file. What will this C file be used for?

sameo replied:
We now generate a blob into a file, or an HJSON OTP image partition to be used as an OTP image overlay. I hope it makes more sense.

sameo commented:
> Thanks for this. 

Thanks for the review.

> The documentation/specification changes look good, modulo a few corrections and clarifications that I've suggested. May I suggest, given the time frames involved, that it would be sensible to separate these from the patch generation tool so that the documentation and specification changes may be integrated sooner?
> 

Yes, that makes sense. I have now extracted and removed the specification changes from this PR into #21070 so that we can merge it first.

> The lowRISC software team will want to review the tool and Rust code thoroughly and I think our priority should be to get the ROMs and e2e testing completed first.

I agree. There are quite a lot of pending comments on the rust tool in this PR, I will address them very shortly.

sameo commented on hw/ip/rv_core_ibex/doc/boot-rom-patching.md:
```diff
@@ -140,7 +140,7 @@ The second ROM partition with patching capability is added to the architecture d
 - **The patch table** includes:
 
-  - Contents for 32 [*patch match register pairs*](#patch-match--redirect-logic), each register 1 DWORD in size.
+  - Contents for 32 [*patch match descriptor pairs*](#patch-match--redirect-logic), each descriptor is 1 DWORD in size.
```
Closing all those comments as the related documentation changes have been merged in a separate PR.
Thanks for the comments @alees24.

sameo commented on sw/host/opentitanlib/src/rom_patch/patch.rs:
```diff
@@ -0,0 +1,745 @@
+            + (N_MATCH_DESCRIPTORS * mem::size_of::<RomPatchMatchDescriptor>())
+            + code_len
+            + RSA_SIGNATURE_LEN_BYTES;
+
```
Good point. I added a check for that.

sameo commented:
@pamaury @engdoreis @jwnrt I finally found some bandwidth to fixed that PR. The forcing function was the ECDSA enablement in `integrated_dev`...

Anyways, this updated PR addresses all pending comments, hopefully. Using the `zerocopy` crate definitely makes the code smaller and there are no unsafe blocks anymore.

As I also wanted to switch to `clap` for the CLI and to ECDSA for the ROM patch signature, this PR carries some commits from `master`.

Thanks in advance for giving that PR a new round of reviews.

engdoreis commented on sw/host/opentitanlib/src/rom_patch/mod.rs:
```diff
@@ -0,0 +1,691 @@
+        let mut code_size_words = regions.iter().map(|r| r.patch_code.len()).sum();
+
+        // Compute the overall code size for the patch.
+        for region in regions {
+            if let Some(sub_routines_hjson) = &region.sub_routines {
+                for s in sub_routines_hjson {
+                    code_size_words =
+                        std::cmp::max(code_size_words, (s.offset >> 2) + s.code.len());
+                }
+            }
+        }
```
You could remove the Option from `sub_routines` combined with `#[serde(default)]` to make it an empty vector in case it is not defined in the hjson, so that the `if let` would not be required. 
Iterator functions could also be used to make the code more idiomatic.
```suggestion
let code_size_words = regions
        .iter()
        .flat_map(|region| region.sub_routines.iter())
        .map(|s| (s.offset >> 2) + s.code.len())
        .fold(code_size_words, std::cmp::max);
```

sameo replied:
Nice. Applied, thanks.

engdoreis reviewed (commented):
Thanks @sameo for the PR.

I've reviewed most of it and let some comments.
I'm going to finish the review of the last commit later.

jesultra reviewed (approved):
I do not know about "ROM patches", but the switch to clap for command line parsing, and general structure of the code look fine to me.

sameo commented on sw/host/opentitanlib/src/otp/otp_img.rs:
```diff
@@ -31,6 +33,19 @@ where
+    let mut seq = serializer.serialize_seq(Some(strings.len()))?;
+    for s in &strings {
+        seq.serialize_element(s)?;
+    }
```
Thanks, suggestion applied now.

sameo commented on sw/host/opentitanlib/src/rom_patch/mod.rs:
```diff
@@ -0,0 +1,691 @@
+            1 << PATCH_LOCKED_FIELD_SHIFT
+        } else {
+            0
+        };
```
Applied as well, thanks.

sameo commented on sw/host/opentitanlib/src/rom_patch/mod.rs:
```diff
@@ -0,0 +1,691 @@
+        };
+
+        let match_address = match match_addr {
+            addr if addr <= PATCH_MATCH_ADDR_FIELD_MASK => addr & PATCH_MATCH_ADDR_FIELD_MASK,
```
Indeed. Fixed now.

sameo commented on sw/host/opentitanlib/src/rom_patch/mod.rs:
```diff
@@ -0,0 +1,691 @@
+#[derive(AsBytes, Copy, Clone, Debug, Default)]
+#[repr(packed, C)]
+struct RomPatchMatchTable {
+    table: [RomPatchMatchDescriptor; N_MATCH_DESCRIPTORS],
```
N_MATCH_DESCRIPTORS is an RTL constant for Darjeeling. But I guess using const generics make it more flexible. I switched the implementation to it now, and although rom_patch.rs uses the exported N_MATCH_DESCRIPTORS, this could become a CLI option in the future.

sameo commented:
> Thanks @sameo for the PR.
> 
> I've reviewed most of it and let some comments. I'm going to finish the review of the last commit later.

Thanks for the review @engdoreis . I think I addressed all your comments now.

engdoreis commented on sw/host/opentitanlib/src/rom_patch/mod.rs:
```diff
@@ -0,0 +1,700 @@
+        unsafe {
+            let ptr = self.code.as_ptr();
+            let len = self.code.len() * std::mem::size_of::<u32>();
+            std::slice::from_raw_parts(ptr as *const u8, len)
```
Only this line needs to be unsafe.

sameo replied:
That's right. Fixed now.

engdoreis commented on sw/host/opentitantool/src/command/rom_patch.rs:
```diff
@@ -0,0 +1,106 @@
+
+        // let json = serialize(&otp_img)?.to_hjson().to_string();
+
+        // if let Some(hjson_output) = &self.hjson_output {
```
Remove commented out code?

sameo replied:
Oops. Fixed as well, thanks.

engdoreis reviewed (approved):
Thanks @sameo for the fixes, besides some other comments below it LGTM.

sameo commented on sw/host/opentitanlib/src/rom_patch/mod.rs:
```diff
@@ -0,0 +1,700 @@
+        let mut code_index = 0;
+
+        // Copy the patch code itself.
+        for region in regions {
```
Sure. Done.

sameo commented:
@engdoreis @jesultra Please let me know if this PR needs some more work.

jesultra commented:
I do not think I can say much more than I already have.  I am not familiar with the feature being built here, but the refactorings in opentitanlib and code overall looks good.  I have not been able to test whether the changes cause issues for `opentitansession` as used at Google, because it seems the branch `integrated_dev` is quite far removed from `master`.  Are these refactorings cherry-picked from `master`, or is there a plan to push the other way?

andreaskurth commented:
This seems to have gotten plenty of reviews (and it has been open for quite some time), and it looks like all has converged. Big congrats and thanks to everyone involved in this effort!
